staging/vhost-device-gpu/src/device.rs

```diff
@@ -0,0 +1,1607 @@
+    DescriptorWriteFailed,
+    #[error("Invalid command type {0}")]
+    InvalidCommandType(u32),
+    #[error("Failed to send used queue notification: {0}")]
+    NotificationFailed(io::Error),
+    #[error("Failed to create new EventFd")]
+    EventFdFailed,
+    #[error("Failed to create an iterator over a descriptor chain: {0}")]
+    CreateIteratorDescChain(virtio_queue::Error),
+    #[error("Failed to create descriptor chain Reader: {0}")]
+    CreateReader(virtio_queue::Error),
+    #[error("Failed to create descriptor chain Writer: {0}")]
+    CreateWriter(virtio_queue::Error),
+    #[error("Failed to decode gpu command: {0}")]
+    GpuCommandDecode(GpuCommandDecodeError),
+    #[error("Failed to encode gpu response: {0}")]
+    GpuResponseEncode(GpuResponseEncodeError),
+    #[error("Failed add used chain to queue: {0}")]
+    QueueAddUsed(virtio_queue::Error),
+    #[error("Epoll handler not available: {0}")]
+    EpollHandler(String),
+    #[error("Failed register epoll listener: {0}")]
+    RegisterEpollListener(io::Error),
+}
+
+impl From<Error> for io::Error {
+    fn from(e: Error) -> Self {
+        Self::new(io::ErrorKind::Other, e)
+    }
+}
+
+struct VhostUserGpuBackendInner {
+    virtio_cfg: VirtioGpuConfig,
+    event_idx_enabled: bool,
+    gpu_backend: Option<GpuBackend>,
+    pub exit_event: EventFd,
+    mem: Option<GuestMemoryAtomic<GuestMemoryMmap>>,
+    gpu_mode: GpuMode,
+}
+
+pub struct VhostUserGpuBackend {
+    inner: Mutex<VhostUserGpuBackendInner>,
+    // this uses sync::Weak to avoid a reference cycle
+    epoll_handler: Mutex<sync::Weak<VringEpollHandler<Arc<Self>>>>,
+    poll_event_fd: Mutex<Option<EventFd>>,
+}
+
+impl VhostUserGpuBackend {
+    pub fn new(gpu_config: &GpuConfig) -> Result<Arc<Self>> {
+        log::trace!("VhostUserGpuBackend::new(config = {:?})", &gpu_config);
+        let inner = VhostUserGpuBackendInner {
+            virtio_cfg: VirtioGpuConfig {
+                events_read: 0.into(),
+                events_clear: 0.into(),
+                num_scanouts: Le32::from(VIRTIO_GPU_MAX_SCANOUTS),
+                num_capsets: RutabagaVirtioGpu::MAX_NUMBER_OF_CAPSETS.into(),
+            },
+            event_idx_enabled: false,
+            gpu_backend: None,
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
+            mem: None,
+            gpu_mode: gpu_config.gpu_mode(),
+        };
+
+        Ok(Arc::new(Self {
+            inner: Mutex::new(inner),
+            epoll_handler: Mutex::new(sync::Weak::new()),
+            poll_event_fd: Mutex::new(None),
+        }))
+    }
+
+    pub fn set_epoll_handler(&self, epoll_handlers: &[Arc<VringEpollHandler<Arc<Self>>>]) {
+        // We only expect 1 thread to which we want to register all handlers
+        assert_eq!(epoll_handlers.len(), 1);
+        let mut handler = match self.epoll_handler.lock() {
+            Ok(h) => h,
+            Err(poisoned) => poisoned.into_inner(),
+        };
```

Comment:
Question (not review) comment, why is it okay if the mutex was poisoned here? Is the epoll handler guaranteed to be valid even if the thread has panicked?

Reply:
Honestly, I am not sure why I did it this way. Thinking about it now, I should just unwrap() the value, we can't do anything useful if the thread panics (even if the epoll_hander state was correct).